### PR TITLE
Implements exponential backoff for Mozilla Location Services

### PR DESCRIPTION
**Pull request by crankycoder**
A patch to add exponential backoff for all 400 and 500 series errors from Mozilla Location Service.

~~My local builds seem broken so I'm just going to wait for Travis to do a build for me.~~

edit: My local build only seems to work with Android Studio

### CODE REVIEW

**Comment by crankycoder** [comment]
This PR should address the issues listed in #32 and which were addressed partially in 4c61799

**Comment by crankycoder** [comment]
cc/ @ckolos

**Comment by ale5000-git** [comment]
I don't know the code exactly but I think that in this way there will be cases where an app request the location in the middle of the request and it will get an updated lastRequestTime but an old lastResponse.

I think it would be better to leave it in the previous place but maybe add another variable to track when a request is in progress and do not allow another one at the same time (if this is the problem).

**Comment by ale5000-git** [comment]
I think there should be an upper limit to RATE_LIMIT_MS to avoid make it grown too much (or overflow) when there are server problems.
And also a lower limit, I don't see it in the latest code change.

**Comment by crankycoder** [comment]
@ale5000-git this adds a floor value and a limit of 1024 times the floor value to set the ceiling backoff time.

I've also shuffled around a lot of the shared variables and protected them with synchronized methods and simplified the setup for a replayed location fix.

**Comment by crankycoder** [comment]
Can we get a review?

**Comment by crankycoder** [comment]
thanks @mar-v-in !  I think I've addressed all your review. 

Regarding https://github.com/microg/IchnaeaNlpBackend/pull/39#pullrequestreview-128060110 - I've reworked the delay calculation for the back off timer to have a hard floor rate of 60sec.

File: src/main/java/org/microg/nlp/backend/ichnaea/BackendService.java [hunk 1]
```diff
@@ -44,10 +44,11 @@
     private static final String TAG = "IchnaeaBackendService";
     private static final String SERVICE_URL = "https://location.services.mozilla.com/v1/geolocate?key=%s";
     private static final String API_KEY = "068ab754-c06b-473d-a1e5-60e7b1a2eb77";
-    private static final long RATE_LIMIT_MS_FLOOR = 10000;
+    private static final long RATE_LIMIT_MS_FLOOR = 60000;
+    private static final long RATE_LIMIT_MS_PADDING = 10000;
     private static final String PROVIDER = "ichnaea";
 
-    private long EXP_BACKOFF_RATE = 1;
+    private long EXP_BACKOFF_FACTOR = 0;
```
Comment (ale5000-git) [inline src/main/java/org/microg/nlp/backend/ichnaea/BackendService.java]: EXP_BACKOFF_FACTOR isn't a constant so as mar-v-in (lowercase starting camel case).

Just a personal note ( I always think about phones with low memory ;) ), does it really need to be long for just 1024 values?
Similare note for RATE_LIMIT_MS_FLOOR and RATE_LIMIT_MS_PADDING.

**Comment by crankycoder** [comment]
You're not saving any memory materially by doing that.  It's going to be a
rounding errors compared to whatever the GC is doing to the heap.

Can we cut to the chase and get this merged in?

MicroG is currently consuming 10x the traffic of all other traffic combined
on the MLS servers because of the query rate and the lack of backoff
throttling is degrading the service for all users.

**Comment by ale5000-git** [comment]
Only mar-v-in can merge it.

**Pull request merged**